### PR TITLE
CNV-33890: Swap feature in CNV RN 4.16

### PR DESCRIPTION
**Pull request by sbeskin-redhat**
Resolves: https://issues.redhat.com/browse/CNV-33890

OCP 4.16
CNV 4.16

Preview: https://77535--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/release_notes/virt-4-16-release-notes.html#virt-4-16-technology-preview

### CODE REVIEW

File: modules/virt-using-wasp-agent-to-configure-higher-vm-workload-density.adoc [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+= Using wasp-agent to configure higher VM workload density
+
+wasp-agent is a component to enable an OpenShift cluster to assign SWAP resources to burstable VM workloads only. See https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/
+SWAP usage is supported on worker nodes only.
+
+.Prerequisites
+
+* The `oc` tool is available.
+* Logged into cluster with cluster-admin role
+* A defined memory over-commit ratio. By default: 150%
```
Comment (fabiand) [inline modules/virt-using-wasp-agent-to-configure-higher-vm-workload-density.adoc:16]: Let's drop this buellet here, a reader could wonder how they get to this value.

File: modules/virt-using-wasp-agent-to-configure-higher-vm-workload-density.adoc [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="virt-using-wasp-agent-to-configure-higher-vm-workload-density_{context}"]
+= Using wasp-agent to configure higher VM workload density
+
+wasp-agent is a component to enable an OpenShift cluster to assign SWAP resources to burstable VM workloads only. See https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/
+SWAP usage is supported on worker nodes only.
```
Comment (fabiand) [inline modules/virt-using-wasp-agent-to-configure-higher-vm-workload-density.adoc:10]: `See kube link` should we have this here?
Rather point to the OCP module?

File: virt/release_notes/virt-4-16-release-notes.adoc [hunk 3]
```diff
@@ -140,9 +140,9 @@ link:https://access.redhat.com/support/offerings/techpreview[Technology Preview
 //CNV-33125: Add CPU limits to the UI
 * Cluster admins can now enable CPU resource limits on a namespace in the {product-title} web console under *Overview* -> *Settings* -> *Preview features*.
 
-[id="virt-4-16-bug-fixes"]
-== Bug fixes
-
+//CNV-22314: Safe memory overcommitment using `wasp-agent`
```
Comment (fabiand) [inline virt/release_notes/virt-4-16-release-notes.adoc:143]: Should this mention that it is in TP?

File: virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+include::_attributes/common-attributes.adoc[]
+:context: virt-configuring-higher-vm-workload-density
+
+toc::[]
+
+
+:FeatureName: Configuring higher workload density
+include::snippets/technology-preview.adoc[]
+
+To save resources and to increase the number of VMs, You can configure a higher VM workload density in your cluster by over-committing memory resources (RAM).
```
Comment (fabiand) [inline virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc:13]: You should be you?

Comment (fabiand) [reply virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc:13]: Lowercase

File: virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+While over-committed memory can lead to a higher workload density, at the same time this will lead to some side-effects:
+Lower workload performance on a highly utilized system
```
Comment (fabiand) [inline virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc:15]: Should this be a bullet list?

File: virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc (hunk 4 above)
Comment (fabiand) [inline virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc:13]: Rewrite the sentence to:
To increase the number of VMs per node, you can configure a higher VM workload density in your cluster by over-committing memory resources (RAM) with swap.

**Review by fabiand — CHANGES_REQUESTED** [review]

File: virt/release_notes/virt-4-16-release-notes.adoc (hunk 3 above)
Comment (sbeskin-redhat) [reply virt/release_notes/virt-4-16-release-notes.adoc:151]: @fabiand  This is in the 'Technology Preview features" section.

**Comment by sbeskin-redhat** [comment]
@fabiand 
Sorry, I forgot to mention that this PR is only about the release notes: virt/release_notes/virt-4-16-release-notes.adoc
The other files relating to the swap feature (virt-configuring-higher-vm-workload-density.adoc and virt-using-wasp-agent-to-configure-higher-vm-workload-density.adoc) are updated in another PR: https://github.com/openshift/openshift-docs/pull/76344

**Comment by max-cx** [comment]
@sbeskin-redhat, :slightly_smiling_face: I'm looking at the [preview](https://77535--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/virtual_machines/virt-configuring-higher-vm-workload-density.html) and it seems that you already added the prerequisites but [still need to add a procedure](https://github.com/openshift/openshift-docs/pull/77535/files#diff-6c4e57708f8c064606b3df0f6729426f324069abe615aeddf390c84d32b2b92dR22). Also, I see the previous reviewer's comments that have not been addressed or responded to. I also recommend consulting the repo rules [files](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/) such as [how to document TP features](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#indicating-technology-preview-features).

:slightly_smiling_face: AFAIK, we normally get peer reviews when we feel our authoring work on the PR is completed, so I suggest that you add the missing procedure, get this PR to a point where you believe that it's ready to be merged, and ask the QE to review and approve it. Also, please remember to squash the commits. And then add the `peer-review-needed` label.

Since I'm only onboarding for the peer review squad, let's ask for a second opinion or additional feedback from @dfitzmau.

**Comment by max-cx** [comment]
> Sorry, I forgot to mention that this PR is only about the release notes: virt/release_notes/virt-4-16-release-notes.adoc
> The other files relating to the swap feature (virt-configuring-higher-vm-workload-density.adoc and virt-using-wasp-agent-to-configure-higher-vm-workload-density.adoc) are updated in another PR: #76344

@sbeskin-redhat, :slightly_smiling_face: in such a case, please remove the changes that should not be merged with this PR.

**Comment by dfitzmau** [comment]
Hi @sbeskin-redhat . I agree with Max. The commits need to be squashed, non-intended files should be reverted, and the merge conflict(s) need to be addressed.

**Comment by sbeskin-redhat** [comment]
@dfitzmau 
Hi,
Squashed, rebased, resolved

File: _topic_maps/_topic_map.yml [hunk 6]
```diff
@@ -4193,6 +4193,8 @@ Topics:
     File: virt-using-vtpm-devices
   - Name: Managing virtual machines with OpenShift Pipelines
     File: virt-managing-vms-openshift-pipelines
+  - Name: Configuring higher VM workload density
```
Comment (dfitzmau) [inline _topic_maps/_topic_map.yml:4196]: Please follow these steps to determine if a redirect is required:

[link](https://docs.google.com/document/d/1fLMpK4bqthtFlCwA36yeo2cI-pExFpX76VMnSi_In6Q/edit#bookmark=id.xb255koulgqx)

File: virt/release_notes/virt-4-16-release-notes.adoc [hunk 7]
```diff
@@ -161,6 +161,9 @@ link:https://access.redhat.com/support/offerings/techpreview[Technology Preview
 //CNV-33125: Add CPU limits to the UI
 * Cluster admins can now enable CPU resource limits on a namespace in the {product-title} web console under *Overview* -> *Settings* -> *Preview features*.
 
+//CNV-22314: Safe memory overcommitment using `wasp-agent`
+* Cluster admins can now xref:../../virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc#virt-configuring-higher-vm-workload-density[use the `wasp-agent` tool to configure a higher VM workload density] in their clusters by overcommitting the amount of memory (RAM) and assigning swap resources to VM workloads.
```
Comment (dfitzmau) [inline virt/release_notes/virt-4-16-release-notes.adoc:165]: For accessibility reasons I'm not sure inline linking is the way to go. How about step out the link on a new line:

> For more information, see Configuring higher VM workload density.

This way when a customer clicks on the link it opens to a matching title.

Comment (dfitzmau) [inline virt/release_notes/virt-4-16-release-notes.adoc:165]:
```suggestion
* Cluster admins can now xref:../../virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc#virt-configuring-higher-vm-workload-density[use the `wasp-agent` tool to configure a higher VM workload density] in their clusters by overcommitting the amount of memory, in RAM, and assigning swap resources to VM workloads.
```

**Review by dfitzmau — APPROVED** [review]
/label peer-review-done

File: _topic_maps/_topic_map.yml (hunk 6 above)
Comment (sbeskin-redhat) [reply _topic_maps/_topic_map.yml:4196]: @dfitzmau  But these are my new files that I created in another PR. I just noticed that I had placed them not where I wanted, by mistake.

Comment (dfitzmau) [reply _topic_maps/_topic_map.yml:4196]: Ah, OK. You are the master of their destiny so I guess you won't trip up anything here.

File: virt/release_notes/virt-4-16-release-notes.adoc (hunk 7 above)
Comment (dfitzmau) [reply virt/release_notes/virt-4-16-release-notes.adoc:165]: I'll leave it to you, but for accessibility reasons, inline links can be tricky.

Comment (sbeskin-redhat) [reply virt/release_notes/virt-4-16-release-notes.adoc]: @dfitzmau  I've left the link inline, but slightly modified the link text to better correspond to the heading of the section it leads to.

**Review by sheriff-rh — APPROVED** [review]
Changes LGTM, merging.

**Pull request merged**